File: cougar_rgb/daemon.py
"""Фоновая служба: держит эффект применённым и рисует программные анимации."""
import fcntl
import logging
import os
import signal
import subprocess
import threading
import time
from pathlib import Path

from . import audio
from .config import CONFIG_PATH, Config
from .device import DeviceNotFound, Fusion2
from .engine import Engine

log = logging.getLogger('cougar-rgb')

LOCK_PATH = Path(os.environ.get('XDG_RUNTIME_DIR', '/tmp')) / 'cougar-rgb.lock'
RECONNECT_DELAY = 2.0
HEALTH_INTERVAL = 5.0
RESUME_SETTLE = 2.0


def is_running():
    try:
        with open(LOCK_PATH, 'a') as f:
            fcntl.flock(f, fcntl.LOCK_EX | fcntl.LOCK_NB)
            fcntl.flock(f, fcntl.LOCK_UN)
            return False
    except BlockingIOError:
        return True


def _config_stamp():
    try:
        return CONFIG_PATH.stat().st_mtime_ns
    except FileNotFoundError:
        return None


def _suspend_gap():
    return time.clock_gettime(time.CLOCK_BOOTTIME) - time.clock_gettime(time.CLOCK_MONOTONIC)


SLEEP_MONITOR = ['stdbuf', '-oL', 'gdbus', 'monitor', '--system', '--dest', 'org.freedesktop.login1',
                 '--object-path', '/org/freedesktop/login1']


def is_resume_signal(line):
    return line.rstrip().endswith('.PrepareForSleep (false,)')


class SleepWatcher(threading.Thread):
    """Ловит пробуждение по сигналу logind PrepareForSleep(false).

    Разница CLOCK_BOOTTIME и CLOCK_MONOTONIC не годится как единственный признак:
    при сбитых аппаратных часах ядро почти не учитывает время сна.
    """

    def __init__(self):
        super().__init__(name='sleep-watch', daemon=True)
        self.resumed = threading.Event()
        self._stop = threading.Event()
        self._proc = None

    def run(self):
        while not self._stop.is_set():
            try:
                self._proc = subprocess.Popen(SLEEP_MONITOR, stdout=subprocess.PIPE,
                                              stderr=subprocess.DEVNULL, text=True)
                for line in self._proc.stdout:
                    if is_resume_signal(line):
                        self.resumed.set()
                self._proc.wait()
            except OSError as e:
                log.warning('cannot watch logind sleep signals: %s', e)
            self._stop.wait(5.0)

    def stop(self):
        self._stop.set()
        if self._proc:
            self._proc.terminate()


class Daemon:
    def __init__(self):
        self.running = True
        self.dev = None
        self.engine = None
        self.stamp = object()
        self.gap = _suspend_gap()
        self.last_health = 0.0
        self.sleep_watcher = SleepWatcher()

    def stop(self, *_):
        self.running = False

    def disconnect(self):
        if self.dev:
            self.dev.close()
        self.dev = self.engine = None

    def connect(self):
        try:
            self.dev = Fusion2()
            self.engine = Engine(self.dev)
            self.engine.setup()
            log.info('connected to %s (%s)', self.dev.path, self.dev.info()['name'])
            self.stamp = object()   # принудительно применить конфиг
            return True
        except OSError as e:
            if not isinstance(e, DeviceNotFound):
                log.warning('connect failed: %s', e)
            self.disconnect()
            return False

    def apply_config(self):
        cfg = Config.load()
        self.engine.apply(cfg.effect, cfg.params(), cfg.audio_delays)
        log.info('applied %s, brightness %d%%', cfg.effect, cfg.brightness)

    def step(self):
        gap = _suspend_gap()
        if self.sleep_watcher.resumed.is_set() or gap - self.gap > 1.0:
            log.info('resume from suspend detected, reinitializing controller')
            time.sleep(RESUME_SETTLE)
            self.sleep_watcher.resumed.clear()
            self.disconnect()
            gap = _suspend_gap()
        self.gap = gap

        if self.dev is None and not self.connect():
            time.sleep(RECONNECT_DELAY)
            return

        stamp = _config_stamp()
        if stamp != self.stamp:
            self.stamp = stamp
            self.apply_config()

        now = time.monotonic()
        if self.engine.animated:
            self.engine.tick()
            time.sleep(1 / self.engine.fps)
        else:
            if now - self.last_health > HEALTH_INTERVAL:
                self.last_health = now
                self.dev.info()     # выбросит OSError, если контроллер пропал
            time.sleep(0.1)

    def run(self):
        signal.signal(signal.SIGTERM, self.stop)
        signal.signal(signal.SIGINT, self.stop)
        self.sleep_watcher.start()
        while self.running:
            try:
                self.step()
            except OSError as e:
                log.warning('device error: %s, reconnecting', e)
                self.disconnect()
                time.sleep(RECONNECT_DELAY)
        self.disconnect()
        self.sleep_watcher.stop()
        audio.analyzer.stop()


def main():
    logging.basicConfig(level=logging.INFO, format='%(levelname)s %(message)s')
    lock = open(LOCK_PATH, 'a')
    try:
        fcntl.flock(lock, fcntl.LOCK_EX | fcntl.LOCK_NB)
    except BlockingIOError:
        raise SystemExit('cougar-rgb daemon is already running') from None
    Daemon().run()

File: src/cli/daemon.cpp
// Background daemon: keeps the effect applied and renders software animations.

#include "cli.hpp"

#include "audio.hpp"
#include "clock.hpp"
#include "config.hpp"
#include "device.hpp"
#include "engine.hpp"
#include "paths.hpp"

#include <atomic>
#include <chrono>
#include <csignal>
#include <cstdarg>
#include <cstdio>
#include <ctime>
#include <fcntl.h>
#include <memory>
#include <optional>
#include <sys/file.h>
#include <sys/stat.h>
#include <systemd/sd-bus.h>
#include <thread>
#include <unistd.h>

namespace cougar::cli {

namespace {

constexpr double RECONNECT_DELAY = 2.0;
constexpr double HEALTH_INTERVAL = 5.0;
constexpr double RESUME_SETTLE = 2.0;

std::atomic<bool> g_running{true};

void log(const char *level, const char *fmt, ...)
{
    std::va_list args;
    va_start(args, fmt);
    std::fprintf(stderr, "%s ", level);
    std::vfprintf(stderr, fmt, args);
    std::fputc('\n', stderr);
    va_end(args);
}

void sleep_for(double seconds)
{
    std::this_thread::sleep_for(std::chrono::duration<double>(seconds));
}

double suspend_gap()
{
    timespec boot{}, mono{};
    ::clock_gettime(CLOCK_BOOTTIME, &boot);
    ::clock_gettime(CLOCK_MONOTONIC, &mono);
    return double(boot.tv_sec - mono.tv_sec) + double(boot.tv_nsec - mono.tv_nsec) / 1e9;
}

std::optional<std::timespec> config_stamp()
{
    struct stat st{};
    if (::stat(paths::config_file().c_str(), &st) != 0)
        return std::nullopt;
    return st.st_mtim;
}

bool same_stamp(const std::optional<std::timespec> &a, const std::optional<std::timespec> &b)
{
    if (!a || !b)
        return !a && !b;
    return a->tv_sec == b->tv_sec && a->tv_nsec == b->tv_nsec;
}

// Detects resume from logind's PrepareForSleep(false) signal.
//
// The CLOCK_BOOTTIME/CLOCK_MONOTONIC gap is not reliable on its own:
// with a wrong hardware clock the kernel barely accounts for the time spent asleep.
class SleepWatcher {
public:
    std::atomic<bool> resumed{false};

    void start() { thread_ = std::thread([this] { run(); }); }

    void stop()
    {
        stop_ = true;
        if (thread_.joinable())
            thread_.join();
    }

private:
    static int on_signal(sd_bus_message *m, void *userdata, sd_bus_error *)
    {
        int start = 0;
        if (sd_bus_message_read(m, "b", &start) >= 0 && !start)
            static_cast<SleepWatcher *>(userdata)->resumed = true;
        return 0;
    }

    void run()
    {
        while (!stop_) {
            sd_bus *bus = nullptr;
            if (sd_bus_open_system(&bus) >= 0 &&
                sd_bus_match_signal(bus, nullptr, "org.freedesktop.login1", "/org/freedesktop/login1",
                                    "org.freedesktop.login1.Manager", "PrepareForSleep", on_signal, this) >= 0) {
                while (!stop_) {
                    const int r = sd_bus_process(bus, nullptr);
                    if (r < 0)
                        break;
                    if (r == 0 && sd_bus_wait(bus, 500'000) < 0)
                        break;
                }
            } else {
                log("WARNING", "cannot watch logind sleep signals");
            }
            sd_bus_flush_close_unref(bus);
            for (int i = 0; i < 50 && !stop_; ++i)
                sleep_for(0.1);
        }
    }

    std::atomic<bool> stop_{false};
    std::thread thread_;
};

class Daemon {
public:
    void run()
    {
        sleep_watcher_.start();
        while (g_running) {
            try {
                step();
            } catch (const DeviceError &e) {
                log("WARNING", "device error: %s, reconnecting", e.what());
                disconnect();
                sleep_for(RECONNECT_DELAY);
            }
        }
        disconnect();
        sleep_watcher_.stop();
        analyzer().stop();
    }

private:
    void disconnect()
    {
        engine_.reset();
        dev_.reset();
    }

    bool connect()
    {
        try {
            dev_ = Fusion2::open();
            engine_ = std::make_unique<Engine>(*dev_);
            engine_->setup();
            log("INFO", "connected to %s (%s)", dev_->path().c_str(), dev_->info().name.c_str());
            force_apply_ = true;
            return true;
        } catch (const DeviceNotFound &) {
        } catch (const DeviceError &e) {
            log("WARNING", "connect failed: %s", e.what());
        }
        disconnect();
        return false;
    }

    void apply_config()
    {
        const Config cfg = Config::load(paths::config_file());
        engine_->apply(cfg.effect, cfg.params(), cfg.audio_delays);
        log("INFO", "applied %s, brightness %d%%", cfg.effect.c_str(), cfg.brightness);
    }

    void step()
    {
        double gap = suspend_gap();
        if (sleep_watcher_.resumed || gap - gap_ > 1.0) {
            log("INFO", "resume from suspend detected, reinitializing controller");
            sleep_for(RESUME_SETTLE);
            sleep_watcher_.resumed = false;
            disconnect();
            gap = suspend_gap();
        }
        gap_ = gap;

        if (!dev_ && !connect()) {
            sleep_for(RECONNECT_DELAY);
            return;
        }

        const auto stamp = config_stamp();
        if (force_apply_ || !same_stamp(stamp, stamp_)) {
            force_apply_ = false;
            stamp_ = stamp;
            apply_config();
        }

        const double now = clock::now();
        if (engine_->animated()) {
            engine_->tick();
            next_frame_ = std::max(next_frame_ + 1.0 / engine_->fps(), now);
            sleep_for(next_frame_ - clock::now());
        } else {
            if (now - last_health_ > HEALTH_INTERVAL) {
                last_health_ = now;
                dev_->info();       // throws DeviceError if the controller is gone
            }
            sleep_for(0.1);
        }
    }

    std::unique_ptr<Fusion2> dev_;
    std::unique_ptr<Engine> engine_;
    SleepWatcher sleep_watcher_;
    std::optional<std::timespec> stamp_;
    bool force_apply_ = true;
    double gap_ = suspend_gap();
    double last_health_ = 0;
    double next_frame_ = 0;
};

}  // namespace

int run_daemon()
{
    const auto lock_path = paths::lock_file();
    const int lock = ::open(lock_path.c_str(), O_RDWR | O_CREAT | O_CLOEXEC, 0600);
    if (lock < 0 || ::flock(lock, LOCK_EX | LOCK_NB) != 0) {
        std::fprintf(stderr, "cougar-rgb daemon is already running\n");
        return 1;
    }
    struct sigaction sa{};
    sa.sa_handler = [](int) { g_running = false; };
    ::sigaction(SIGTERM, &sa, nullptr);
    ::sigaction(SIGINT, &sa, nullptr);

    Daemon().run();
    ::close(lock);
    return 0;
}

}  // namespace cougar::cli
